### ISSUE TITLE
the question of ncclAllreduce duration time in NSight System

### ISSUE THREAD

**Issue by wuyujiji** [NVIDIA/nccl#470]
Excuse me! Recently I used `nsys` to dump inter-node and intra-node's message file in  BERT Large experiments of `16 GPUs 2 machines`, and surprisingly found the `ncclAllreduce` duration time of `intra-node` is longer than `inter-node` **at the same step**. Based on my current knowledge, I understand the the communication delay of `inter-node` is larger than `intra-node`, and thus the ncclAllreduce` duration time of` inter-node` should be longer than `intra-node`. However, the duration time of `nsys` qdrep file exactly the opposite.

**Comment by sjeaugey** [comment]
That's indeed surprising. Can you reproduce that problem running the [NCCL perf tests](https://github.com/nvidia/nccl-tests) ? That would give us an idea of the performance you are getting intra- and inter- node, as well as the times we're considering here.

**Comment by wuyujiji** [comment]
Nccl test also reproduce the problem, as shown follows:

**shell command:**

`LD_LIBRARY_PATH=/sources/nccl278/build/lib NCCL_DEBUG=INFO mpirun --hostfile hostfile   
--prefix /usr/local/ompi -bind-to none 
-map-by slot 
--display-map --tag-output --timestamp-output 
--mca pml ob1 --mca btl_vader_single_copy_mechanism none --mca btl_openib_cpc_include rdmacm --mca btl_openib_rroce_enable 1 --mca btl_tcp_if_exclude lo,docker0 --mca orte_base_help_aggregate 0 --mca btl_openib_receive_queues P,256,256::S,128,256,192,128:S,2048,1024,1008,64:S,12288,1024,1008,64:S,131072,1024,1008,64 
--mca btl tcp,self,vader 
-x NCCL_SOCKET_IFNAME=br0 -x NCCL_ALGO=Ring -x NCCL_DEBUG_SUBSYS=INIT,P2P,GRAPH,ENV -x NCCL_NET_GDR_LEVEL=0 -x NCCL_IB_GID_INDEX=3 -x NCCL_IB_DISABLE=0 -x HOROVOD_MPI_THREADS_DISABLE=1 -x PATH -x PYTHONPATH -x LD_LIBRARY_PATH /opt/nvidia/nsight-systems-cli/2020.3.1/bin/nsys profile -o /workdir/nccl_test.%h.%q{OMPI_COMM_WORLD_RANK} 
./nccl-2.5.7/nccl-tests/build/all_reduce_perf -b 8 -e 512M -f 2 -g 1`

**nccl debug info:**

- nccl channel：
![image](https://user-images.githubusercontent.com/25579435/109407586-f5d32480-79bc-11eb-92df-45441a221498.png)

- bandWidth results:
![image](https://user-images.githubusercontent.com/25579435/109407603-1e5b1e80-79bd-11eb-8eab-0176878e9e27.png)

**nsys results (sorted by duration time):**

- inter-nodes(rank 0 and 1)：

![image](https://user-images.githubusercontent.com/25579435/109407803-ef45ac80-79be-11eb-98d5-705ad594ee01.png)

![image](https://user-images.githubusercontent.com/25579435/109407807-f79de780-79be-11eb-8532-766775ac10ef.png)

- intra-nodes(rank4 and 6):

![image](https://user-images.githubusercontent.com/25579435/109407823-08e6f400-79bf-11eb-94b6-16826ef906b6.png)

![image](https://user-images.githubusercontent.com/25579435/109407825-0edcd500-79bf-11eb-9d34-692aa915d1e0.png)

**Comment by wuyujiji** [comment]
Nccl test also reproduce the problem, as shown follows:

**shell command:**

`LD_LIBRARY_PATH=/sources/nccl278/build/lib NCCL_DEBUG=INFO mpirun --hostfile hostfile   
--prefix /usr/local/ompi -bind-to none 
-map-by slot 
--display-map --tag-output --timestamp-output 
--mca pml ob1 --mca btl_vader_single_copy_mechanism none --mca btl_openib_cpc_include rdmacm --mca btl_openib_rroce_enable 1 --mca btl_tcp_if_exclude lo,docker0 --mca orte_base_help_aggregate 0 --mca btl_openib_receive_queues P,256,256::S,128,256,192,128:S,2048,1024,1008,64:S,12288,1024,1008,64:S,131072,1024,1008,64 
--mca btl tcp,self,vader 
-x NCCL_SOCKET_IFNAME=br0 -x NCCL_ALGO=Ring -x NCCL_DEBUG_SUBSYS=INIT,P2P,GRAPH,ENV -x NCCL_NET_GDR_LEVEL=0 -x NCCL_IB_GID_INDEX=3 -x NCCL_IB_DISABLE=0 -x HOROVOD_MPI_THREADS_DISABLE=1 -x PATH -x PYTHONPATH -x LD_LIBRARY_PATH /opt/nvidia/nsight-systems-cli/2020.3.1/bin/nsys profile -o /workdir/nccl_test.%h.%q{OMPI_COMM_WORLD_RANK} 
./nccl-2.5.7/nccl-tests/build/all_reduce_perf -b 8 -e 512M -f 2 -g 1`

**nccl debug info:**

- nccl channel：
![image](https://user-images.githubusercontent.com/25579435/109407586-f5d32480-79bc-11eb-92df-45441a221498.png)

- bandWidth results:
![image](https://user-images.githubusercontent.com/25579435/109407603-1e5b1e80-79bd-11eb-8eab-0176878e9e27.png)

**nsys results (sorted by duration time):**

- inter-nodes(rank 0 and 1)：

![image](https://user-images.githubusercontent.com/25579435/109407803-ef45ac80-79be-11eb-98d5-705ad594ee01.png)

![image](https://user-images.githubusercontent.com/25579435/109407807-f79de780-79be-11eb-8532-766775ac10ef.png)

- intra-nodes(rank4 and 6):

![image](https://user-images.githubusercontent.com/25579435/109407823-08e6f400-79bf-11eb-94b6-16826ef906b6.png)

![image](https://user-images.githubusercontent.com/25579435/109407825-0edcd500-79bf-11eb-9d34-692aa915d1e0.png)

**Comment by sjeaugey** [comment]
I'm not sure I understand the problem... All GPUs take the same amount of time (92 ms) for allreduce, which is expected, as intra-node / inter-node doesn't matter in a global communicator, all ranks exit the operation at the same time and the slowest links determine the global time of the collective operation.

If your concern is with the first iteration "1", then it's simply that ranks 0/1 probably reach the first allreduce later than the others (they may do more work before), so when they finally reach allreduce, all ranks have been waiting for them for some time. The operation tales 92ms as usual, but for those which reached the allreduce in advance (e.g. 30ms in advance for rank 4), the operation appears to take longer for them. So this is just GPU de-sync and it's not related to how GPUs communicate with others.

**Comment by wuyujiji** [comment]
Okay! In other words, the ncclAllreduce operation of intra-rank (reached the allreduce in advance) in Nsight System contrains two stage: **wait for other ranks** and **do allreduce when other ranks reached**. Do I understand correctly? If correct, How can I only obtain the second stage time (namely do allreduce when other ranks reached)?

**Comment by sjeaugey** [comment]
Indeed, allreduce is a synchronous operation, which will wait until all ranks reach the operation to be able to complete, consuming GPU computing resources. That said, looking at your workload, aside from the first iteration which varies a bit across ranks, all the others seem to show a very stable 92-93ms, which is very likely the real communication+reduction time. Hence there is no reason to suspect blocking consumes significant amounts of computing resources and things seem to be running the best they can.

And again, intra-node/inter-node is not an important consideration here. The operation is not done in two stages but through a pipeline. All ranks will roughly complete at the same time; they're all synchronized at every step of the allreduce. Sure, on 2 nodes we use the tree algorithm which is a bit more de-synchronized than ring, but overall if all ranks do the same amount of work between the different allreduce operations, it is as optimized as it can be.

**Comment by jain-jainendra** [comment]
@sjeaugey How to find reduction (computation) time from total time of Allreduce?

"That said, looking at your workload, aside from the first iteration which varies a bit across ranks, all the others seem to show a very stable 92-93ms, which is very likely the real communication+reduction time."

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
